pandas/io/parsers.py

```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
     return nrows
 
 
+_compression_to_extension = {
+    'gzip': '.gz',
+    'bz2': '.bz2',
+    'zip': '.zip',
+    'xz': '.xz',
+}
+
+
+def _infer_compression(filepath_or_buffer):
+    """
+    Infer compression of a filepath or buffer. In case of buffer, compression
+    is None. Otherwise, inference is perfomed using the extension of the
+    filename or URL.
+    """
+    if not isinstance(filepath_or_buffer, compat.string_types):
+        return None
+    filepath = str(filepath_or_buffer)
+    for compression, extension in _compression_to_extension.items():
+        if filepath.endswith(extension):
+            return compression
+    return None
```

Comment:
I would make this raise a nice ValueError if you have an *invalid* compress (accepting ``None/infer`` as valid). , then we can just call this.